clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -860,65 +870,9 @@ class KernelObjVisitor {
                    const CXXRecordDecl *Wrapper, Handlers &... handlers);
-
-  // Base case, only calls these when filtered.
-  template <typename... FilteredHandlers, typename ParentTy>
-  std::enable_if_t<(sizeof...(FilteredHandlers) > 0)>
-  VisitUnion(FilteredHandlers &... handlers, const CXXRecordDecl *Owner,
-             ParentTy &Parent, const CXXRecordDecl *Wrapper) {
-    (void)std::initializer_list<int>{
-        (handlers.enterUnion(Owner, Parent), 0)...};
-    VisitRecordHelper(Wrapper, Wrapper->fields(), handlers...);
-    (void)std::initializer_list<int>{
-        (handlers.leaveUnion(Owner, Parent), 0)...};
-  }
-
-  // Handle empty base case.
-  template <typename ParentTy>
+  template <typename ParentTy, typename... Handlers>
   void VisitUnion(const CXXRecordDecl *Owner, ParentTy &Parent,
-                  const CXXRecordDecl *Wrapper) {}
-
-  template <typename... FilteredHandlers, typename ParentTy,
-            typename CurHandler, typename... Handlers>
-  std::enable_if_t<!CurHandler::VisitUnionBody &&
-                   (sizeof...(FilteredHandlers) > 0)>
-  VisitUnion(FilteredHandlers &... filtered_handlers,
-             const CXXRecordDecl *Owner, ParentTy &Parent,
-             const CXXRecordDecl *Wrapper, CurHandler &cur_handler,
-             Handlers &... handlers) {
-    VisitUnion<FilteredHandlers...>(filtered_handlers..., Owner, Parent,
-                                    Wrapper, handlers...);
-  }
-
-  template <typename... FilteredHandlers, typename ParentTy,
-            typename CurHandler, typename... Handlers>
-  std::enable_if_t<CurHandler::VisitUnionBody &&
-                   (sizeof...(FilteredHandlers) > 0)>
-  VisitUnion(FilteredHandlers &... filtered_handlers,
-             const CXXRecordDecl *Owner, ParentTy &Parent,
-             const CXXRecordDecl *Wrapper, CurHandler &cur_handler,
-             Handlers &... handlers) {
-    VisitUnion<FilteredHandlers..., CurHandler>(
-        filtered_handlers..., cur_handler, Owner, Parent, Wrapper, handlers...);
-  }
-
-  // Add overloads without having filtered-handlers
-  // to handle leading-empty argument packs.
-  template <typename ParentTy, typename CurHandler, typename... Handlers>
-  std::enable_if_t<!CurHandler::VisitUnionBody>
-  VisitUnion(const CXXRecordDecl *Owner, ParentTy &Parent,
-             const CXXRecordDecl *Wrapper, CurHandler &cur_handler,
-             Handlers &... handlers) {
-    VisitUnion(Owner, Parent, Wrapper, handlers...);
-  }
-
-  template <typename ParentTy, typename CurHandler, typename... Handlers>
-  std::enable_if_t<CurHandler::VisitUnionBody>
-  VisitUnion(const CXXRecordDecl *Owner, ParentTy &Parent,
-             const CXXRecordDecl *Wrapper, CurHandler &cur_handler,
-             Handlers &... handlers) {
-    VisitUnion<CurHandler>(cur_handler, Owner, Parent, Wrapper, handlers...);
-  }
+                  const CXXRecordDecl *Wrapper, Handlers &... handlers);
```

Comment:
The implementation needs to be out of line, since it needs access to some other types.